File: P10533.c
// 10533 - Digit Primes
// has to use index for 'end'

#include <stdio.h>
#include <stdlib.h>

#define isComposite_length 1000001
int isComposite[isComposite_length];
int dp[30200];
int dpLen;

void sieve(int n) {
  int i, j;
  dpLen = 0;
  isComposite[0] = 1;
  isComposite[1] = 1;
  for (i = 2; i < isComposite_length; i++) {
    if (!isComposite[i]) {
      int m = i;
      int s = 0;
      while (m != 0) {
        s += m % 10;
        m /= 10;
      }
      if (!isComposite[s]) {
        dp[dpLen++] = i;
      }
      for (j = i + i; j < isComposite_length; j += i) {
        isComposite[j] = 1;
      }
    }
  }
  dp[dpLen] = 2000000000;  // terminate
}

int main() {
  sieve(1000000);
  int t, tt;
  scanf("%d", &tt);
  for (t = 0; t < tt; t++) {
    int start, end;
	scanf("%d", &start);
	scanf("%d", &end);
	int i = 0;
    while (dp[i++] < start);

    int j;
    if (end > 809213) j = dpLen - 1;
    else if (end > 626953) j = 25000;
    else if (end > 453371) j = 20000;
    else if (end > 287851) j = 15000;
    else j = 10000;
    
	while (dp[j--] > end);

    printf("%d\n", j - i+3);
  }
}

File: P10533.java
// 10533 - Digit Primes
// TLE 

import java.util.Scanner;

public class Main {

  final static Scanner sc = new Scanner(System.in);

  static boolean[] isComposite;
  static boolean[] isComposite2;

  static void sieve(int n) {
    isComposite = new boolean[n + 1];
    isComposite[0] = true;
    isComposite[1] = true;
    //Arrays.fill(isNotPrime, 0, n, true);
    for (int i = 2; i < isComposite.length; i++) {
      if (!isComposite[i]) {
//        System.out.println(i);
        for (int j = i + i; j < isComposite.length; j += i) {
          isComposite[j] = true;
        }
      }
    }
    isComposite2 = new boolean[n + 1];
    for (int i = 0; i < isComposite.length; i++) {
      if (isComposite[i])
        isComposite2[i] = true;
      else {
        isComposite2[i] = isComposite[sum(i)];
      }
    }
  }

  static int sum(int n) {
    int s = 0;
    while (n != 0) {
      s += n % 10;
      n /= 10;
    }
    return s;
  }

  public static void main(String[] args) {
    sieve(1000000);
//    System.out.println("isComposite = " + java.util.Arrays.toString(isComposite));
    int tt = sc.nextInt();
    for (int t = 0; t < tt; t++) {
      int start = sc.nextInt();
      int end = sc.nextInt();
      int count = 0;
      for (int i = start; i <= end; i++) {
        if (!isComposite2[i]) {
          count++;
        }
      }
      System.out.println(count);
    }
  }
}
